Integrate Piece Tree -> IncrementalTextDocument | TextDocuments Incremental Document Support

Jma353 opened the pull request:
This pull request is the result of a dialogue in #476.

This PR does 2 things: 
1. Integrates @rebornix's [`vscode-piece-tree`](https://github.com/rebornix/PieceTree) `npm` module.  I use this module to implement `IncrementalTextDocument`, which is the Incremental sync document representation (as `FullTextDocument` is the Full sync document representation).  This is a class internal to `server` module.
2. Updates the `TextDocuments` document manager to support `Incremental` text synchronization mode.  This manager defaults to `Full`, but has an optional constructor parameter that allows the consumer of the API to specify `Incremental` document sync.  When `Incremental` document sync is specified, this document manager passes `isFullSync = false` to `TextDocument#create`, which results in `IncrementalTextDocument` being used per document.

This PR also bootstraps a test suite for `TextDocuments` (specifically involving `onChange` events) + a test suite for `IncrementalTextDocument` via `TextDocuments` document manager.

Output of `server` tests: 

```
  Connection Tests
    ✓ Ensure request parameter passing
    ✓ Ensure notification parameter passing

  TextDocuments Tests
    ✓ onDidChangeContent TextDocumentSyncKind.Full change file content
    ✓ onDidChangeContent TextDocumentSyncKind.Full several content updates
    ✓ onDidChangeContent TextDocumentSyncKind.Incremental removing content
    ✓ onDidChangeContent TextDocumentSyncKind.Incremental adding content
    ✓ onDidChangeContent TextDocumentSyncKind.Incremental replacing content
    ✓ onDidChangeContent TextDocumentSyncKind.Incremental several content changes

  IncrementalTextDocument
    ✓ Single line document
    ✓ Multiple line document
    ✓ Varying newline characters
    ✓ getText(Range)
    ✓ Invalid inputs
    ✓ Basic append
    ✓ Multi-line append
    ✓ Basic delete
    ✓ Multi-line delete
    ✓ Single character replace
    ✓ Multi-character replace
    ✓ Invalid update ranges


  20 passing (37ms)
```

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/vscode-languageserver-node?pullRequest=478) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/vscode-languageserver-node?pullRequest=478) before we can accept your contribution.<br/><br/>:x: Jma353 [sign now](https://cla.opensource.microsoft.com/Microsoft/vscode-languageserver-node?pullRequest=478)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/vscode-languageserver-node?pullRequest=478) it.</sub>

rebornix commented on server/src/main.ts:
```diff
@@ -255,17 +258,50 @@ export class TextDocuments {
+					let updatedDoc: UpdateableDocument = edits.reduce(
+						(workingTextDoc: UpdateableDocument, edit) => {
+							let newText = TextDocument.applyEdits(workingTextDoc, [edit]);
+							workingTextDoc.update({ text: newText }, version);
```
Currently every `update` on the TextDocument will flush the `_lineOffset`

https://github.com/Microsoft/vscode-languageserver-node/blob/9de8809400b7c443bc25f211dfaddcb74675b2d7/types/src/main.ts#L2654

while `applyEdits` relies on `_lineOffset` https://github.com/Microsoft/vscode-languageserver-node/blob/9de8809400b7c443bc25f211dfaddcb74675b2d7/types/src/main.ts#L2500-L2501

It means for every content change we try to apply, it will go through the whole document, character by character, to find all line breaks. In real world scenarios it can be really costly.

Jma353 replied:
Yup, I definitely observed this limitation.

These black-box functions (`applyEdits` / `update`) are the only ways to currently update text documents.  As we discussed in #476, I can use your npm module in this PR and swap out the implementations for these core functions as well.

Apologies, when I originally wrote this up, I wasn't aware that you had already published the package to npm.

rebornix replied:
No worries and you are doing a great job here. I'm just a bit surprised to see that `lineOffsets` were being calculated every time when we do full sync (the good part is we calculate only when reading line info).

Based on your proposal and this PR, I think we can see two major performance improvement

* We don't need to send full document from client to server anymore when the sync mode is incremental
* The line offset calculation is incremental when the sync mode is incremental, which is actually the major time spent in loading a file in VS Code.

Let me know if you need any help from me, I can push directly to your PR with piece tree related changes (if you are okay with that).

Jma353 replied:
If you've made some sort of local progress on adding your piece tree as a dependency and using it, then feel free to push it to this PR.  Otherwise, I was planning on adding it as a dependency + using it in `FullTextDocument` / `TextDocument` namespace today.

Feel free to let me know, and we can proceed from there.

rebornix replied:
@Jma353 go for it, I don't have anything yet, look forward to your change.

Jma353 replied:
For sure - I'll update this today or tomorrow.

Jma353 commented on types/src/main.ts:
```diff
@@ -2717,6 +2730,103 @@ class FullTextDocument implements TextDocument {
+
+		// Prepare Piece Tree
+		const ptBuilder = new PieceTreeTextBufferBuilder();
+		ptBuilder.acceptChunk(content);
```
@rebornix - should I be slicing this up into 64 KB chunks, as you mentioned is done by `fs.readFile` in your [blog post](https://code.visualstudio.com/blogs/2018/03/23/text-buffer-reimplementation), or is this an appropriate use of your API (passing the entire `content` "chunk" in on instantiation).

rebornix replied:
If the document comes as a single chunk of string, then we just use that. If it's super large, then the performance of editing at the very beginning is somewhat costly (as checking char code is slow on large strings (100+MB)) but it won't be worse than the original approach.

Jma353 commented:
@dbaeumer given that I add piece tree `npm` module in this PR, unsure if I now need to add a `thirdpartynotices.txt` file in `types`, similar to [the one in `client`](https://github.com/Microsoft/vscode-languageserver-node/blob/master/client/thirdpartynotices.txt).  I'm assuming `vscode-piece-tree` will eventually become first-party (under Microsoft github).

rebornix commented on types/src/main.ts:
```diff
@@ -2611,6 +2616,15 @@ export interface TextDocumentContentChangeEvent {
 	text: string;
```
Should we update the comments for `text`? In FullSync mode, `text` is always the content of the full document from client side but now in incremntal mode, it's the change for the corresponding range. Another thing we need to make sure that Client sends only content change but not full text in incremtnal mode.

rebornix replied:
https://github.com/Microsoft/vscode-languageserver-node/blob/053149a76e7ec147347e41a5847a1cc4ccb3d541/client/src/client.ts#L936-L940

It's already guaranteed that when it's incremental, it sends the change event but not the full doc. So we only need to update the documentation.

Jma353 replied:
Will update the documentation in the next commit to the PR :)

rebornix commented on types/src/main.ts:
```diff
@@ -4,6 +4,10 @@
 'use strict';
 
+import { PieceTreeTextBufferBuilder, PieceTreeBase, DefaultEndOfLine } from 'vscode-piece-tree';
+import { Range as PieceTreeRange } from 'vscode-piece-tree/out/common/range';
```
I should probably rename the `out` folder to `lib` or `api`.

Jma353 replied:
Yea, if you wanna change this + bump the version, I can reinstall + fix.

rebornix replied:
I pushed the fix directly to your PR ;)

Jma353 commented:
Outside of updating documentation, any additional feedback on this.  I can address it all in one commit if there's any more.

rebornix commented:
@Jma353 the code looks good to me and thanks for your help! I'll love to hear feedback from others as well, considering some of them are in vacation, I think you can push changes and even better do some benchmark.

Jma353 commented:
@rebornix I'll submit doc change + will setup a lightweight benchmark this weekend.  

In general, I feel as though an explicit win of this is just the idea that we can support incremental sync-ing in the first place.  Sending full files between processes isn't cheap.

aeschli commented:
My comment is about `TextDocument.create(uri: string, languageId: string, version: number, content: string, isFullSync: boolean = true): TextDocument`

The `isFullSync` looks strange, and the return type is always TextDocument. It's unclear to an API user when to set isFullSync and when not.
What about defining a separate constructor?

When doing that I would also officially introduce the UpdatableTextDocument. The default implementation of a UpdatableTextDocument should be the new piece tree document as it is optimised for updates.

export interface UpdatableTextDocument extends TextDocument {
    update(event: TextDocumentContentChangeEvent, version: number)
}
export namespace UpdatableTextDocument {
    export function create(uri: string, languageId: string, version: number, content: string) : UpdatableTextDocument {
       return new IncrementalTextDocument(uri, languageId, version, content);
   }
}

dbaeumer commented:
I though about this again and I am in favour of making this more explicit. However I don't like the name `UpdatableTextDocument` since `FullTextDocument` are updateable as well. To make this in line with the `TextDocumentSyncKind` I would call them `FullTextDocument` and `IncrementalTextDocument`. We should deprecate the `TextDocument` namespace and expose a FullTextDocument interface and namespace. The `FullTextDocument` interface would define an `update(text: string, version)` method. The `IncrementalTextDocument` text document a `update(event: TextDocumentContentChangeEvent, version: number)` method. We can also add `is` methods to the namespace which do an instanceof against the private class.

@Jma353 are you willing to do these changes?

rebornix commented:
I prefer `FullTextDocument` and `IncrementalTextDocument`, which aligns with `TextDocumentSyncKind`.

Jma353 commented:
These changes seem reasonable.  I appreciate the feedback all!

Apologies for the slowness in my response (been busy with some other things).  I will prioritize iterating on this by EOW.

aeschli commented:
@Jma353 @rebornix  Me and Dirk had some more discussion and came up with one more suggestion, which has the smallest impact on the API:
- we keep `types` as-is, no new APIs
- `IncrementalTextDocument` stays in `protocol` as an internal class to be used by`TextDocuments` and optimized for its needs.

Sorry for the back and forth. If you are very busy then I'm happy to merge and make the changes. Just let me know.

Jma353 commented:
I can change these things @aeschli - thanks for the thoughtful feedback.  

Expect an update with said changes by Friday, this way you don't have to do any additional refinements on your part.

Jma353 commented:
@aeschli did you mean `server` (not `protocol`) in your above feedback? 

`protocol` seems to be only `namespace` and `interface` declarations.  Additionally, `TextDocuments` is in `server` subdirectory.

I can remove changes from `types` directory.  The only API change that will result from this PR is an optional argument passed to `TextDocuments` ctor. I currently pass it as a enum value.

Jma353 commented:
I have updated the PR according to feedback, to limit API changes.  Here is a summary: 

1. As mentioned, the only API change is an optional constructor parameter passed to `TextDocuments`, which allows for enum-based switching between `FullTextDocument` and `IncrementalTextDocument`.  
2. `IncrementalTextDocument` is now a class internal to `server` (since it's only used by `TextDocuments`).  
3. I have ported all test-cases that existed in `types` for `IncrementalTextDocument` to `server`.  I have the suite in the `TextDocuments` test file.  
4. I have rebased on @rebornix 's change to the piece tree npm version, as well as latest master.

rebornix commented:
Considering @Jma353 's great effort is going to be merged into the project, I'll add some scripts to VS Code in June to ship piece tree directly from there instead of having my own repository, similar to how we ship Monaco.

Jma353 commented:
@rebornix as in have piece tree available via `vscode` API suite?  

A.k.a. `import {PieceTreeBase, ...} from 'vscode';`?

rebornix commented:
@Jma353 I didn't go that route finally but aligned with how we ship `vscode-uri`, which has its implementation in its own repo. I published a npm package called `vscode-textbuffer`, the api doesn't change at all and the repo will be under microsoft (will be public soon).

dbaeumer commented:
@Jma353 I am correct that you still need to adopt the new `vscode-textbuffer` npm package. Right?

rebornix commented:
I replaced `vscode-piece-tree` with `vscode-textbuffer` and comments related to that.

Jma353 commented:
@rebornix I see travis failure relating to the new module. 

Thanks for swapping out npm modules btw :)

rebornix commented:
@Jma353 thanks it's green green now.

dbaeumer commented:
@Jma353 & @rebornix thanks a lot for making this happen.
